xarray/core/types.py

```diff
@@ -43,18 +47,15 @@
     except ImportError:
         ZarrArray = np.ndarray
 
-    # TODO: Turn on when https://github.com/python/mypy/issues/11871 is fixed.
-    # Can be uncommented if using pyright though.
-    # import sys
-
-    # try:
-    #     if sys.version_info >= (3, 11):
-    #         from typing import Self
-    #     else:
-    #         from typing_extensions import Self
-    # except ImportError:
-    #     Self: Any = None
-    Self: Any = None
+    import sys
+
+    if sys.version_info >= (3, 11):
+        from typing import Self
```

Comment:
Nice! Finally this is merged.

This means we should be able to remove lots of the code that annotated self with a typevar (in another PR ofc).